### PR TITLE
added files for review

### PR DESCRIPTION
**Pull request by ezbc**
Functioning demo from drug-portal-ner-app:

See https://github.com/ezbc/drug-portal-ner-app/commit/7fef38d0f14a60fa5a84f9d1c37631d305a74e58 for the files added.

### CODE REVIEW

File: main.py [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+PLACEHOLDER_RESPONSE ={'entities': [{'text': 'may', 'start_char': 34, 'end_char': 37, 'label': 'Factor'}, {'text': 'significant', 'start_char': 46, 'end_char': 57, 'label': 'Severity'}, {'text': 'hypokalemia', 'start_char': 58, 'end_char': 69, 'label': 'AdverseReaction'}, {'text': 'decrease in serum potassium', 'start_char': 200, 'end_char': 227, 'label': 'AdverseReaction'}, {'text': 'Beta-agonist medications', 'start_char': 281, 'end_char': 305, 'label': 'DrugClass'}, {'text': 'may', 'start_char': 306, 'end_char': 309, 'label': 'Factor'}, {'text': 'transient', 'start_char': 318, 'end_char': 327, 'label': 'Severity'}, {'text': 'hyperglycemia', 'start_char': 328, 'end_char': 341, 'label': 'AdverseReaction'}, {'text': 'potassium', 'start_char': 500, 'end_char': 509, 'label': 'AdverseReaction'}]} 
+
+
+def _get_entities_from_text(request):
+
+    if request.method == 'POST':
+        input_text = request.form.get('text')
+        if input_text == PLACEHOLDER_TEXT:
```
Comment (ezbc) [inline main.py]: The placeholder text response should not be mocked in the production application. Instead we should configure the development server to return the mocked response.

Remove the logic to mock the placeholder text response.

File: main.py [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            text = input_text
+    elif request.method == 'GET':
+        entities = None
+        text = PLACEHOLDER_TEXT
+
+    return entities, text
+
+
+def _annotate_text(annotations, text):
+
```
Comment (ezbc) [inline main.py]: Add a docstring

File: main.py [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+    locs = np.empty((len(annotations), 2), dtype=np.int16)
+    for i, annotation in enumerate(annotations):
+        locs[i, :] = (annotation['start_char'], annotation['end_char'])
+    locs = np.sort(locs, axis=0)
+
+    annotations.sort(key=lambda x: x['start_char'])
+
+    def build_annotation_block(text, label):
```
Comment (ezbc) [inline main.py]: Move this function outside of `_annotate_text`. `build_annotation_block` is a static function that does not need to be instantiated every time `annotate_text` is called.

File: main.py [hunk 4]
```diff
@@ -0,0 +1,171 @@
...
+        else:
+            text_block += char
+            if i == len(text) - 1:
+                block = build_annotation_block(text_block, None)
+                annotated_text.append(block)
+
+    return annotated_text
+
+
+def _base64_decode(encoded_str):
```
Comment (ezbc) [inline main.py]: this function is never called. Remove it.

File: main.py [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    if num_missed_paddings != 4:
+        encoded_str += b'=' * num_missed_paddings
+    return base64.b64decode(encoded_str).decode('utf-8')
+
+
+@app.route('/', methods=['POST', 'GET'])
+def index():
+
+    """Identify FDA adverse events from text"""
+    # TODO: change variable scops in conditionals
```
Comment (ezbc) [inline main.py]: move TODO comment to `_annotate_text`

File: main.py [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+    annotated_text = None
+    if entities is not None:
+        annotated_text = _annotate_text(entities['entities'], text)
+
+    return render_template('index.html',
+                           input=text,
+                           annotated_text=annotated_text)
+
+
+@app.route('/ner/drug', methods=['POST'])
```
Comment (ezbc) [inline main.py]: Remove this route and function, the home page performs the same behavior.

File: requirements.txt [hunk 7]
```diff
@@ -0,0 +1,9 @@
+Flask==0.12.2
```
Comment (ezbc) [inline requirements.txt]: All python dependencies are managed in the docker image. Remove this file.

File: templates/index.html [hunk 8]
```diff
@@ -0,0 +1,170 @@
+{#
```
Comment (ezbc) [inline templates/index.html]: Remove lines 1 - 15. These are irrelevant.

File: templates/index.html [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+
+	  <!-- Latest compiled and minified CSS -->
+		<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
+
+	  <title>Drug Adverse Event Named Entity Recognition</title>
+
+
+    <!-- Custom CSS -->
+    <style>
```
Comment (ezbc) [inline templates/index.html]: Move all css into a static file `static/css/main.css`. Replace with a stylesheet link like the following:
`<link rel="stylesheet" href="{{ url_for('static', filename='css/main.css') }}">`

File: templates/index.html [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#}
+<!DOCTYPE html>
+<html lang="en">
+  <head>
```
Comment (ezbc) [inline templates/index.html]: Create a layout file to slim down templates to just their body content. Each template can extend the layout file to adopt the styling of the app.

## Steps

Move the header including CSS loads and footer including js script loads to a new file called `layout.html`.
In the body of `layout.html` add the following

``` jinja
{% block content %}
{% endblock %}
```

then in all templates extend `layout.html` in the following way:
``` jinja
{% extends "layout.html" %}
{% block content %}
  <!-- body html goes here -->
{% endblock %}
```

For more details visit: http://jonathansoma.com/tutorials/webapps/layouts/

File: templates/index.html [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+    <!-- HTML5 shim and Respond.js for IE8 support of HTML5 elements and media queries -->
+    <!--[if lt IE 9]>
+      <script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
+      <script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
+    <![endif]-->
+  </head>
+
+  <body padding-top="50px">
+
+    <nav class="navbar navbar-inverse navbar-fixed-top">
```
Comment (ezbc) [inline templates/index.html]: Move navbar into a header template. Include the header template in the layout so that all pages will have the navigation header.

File: templates/index.html [hunk 12]
```diff
@@ -0,0 +1,170 @@
...
+  				</form>
+
+          <!-- RESPONSE -->
+          {% include 'response.html' %}
+
+  				<!-- [END form] -->
+  	      </div>
+          <br/>
+
+          <h2>Background</h2>
```
Comment (ezbc) [inline templates/index.html]: Move background to another page. Add a navigation link in the navigation bar.

File: templates/index.html [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+          <br/>
+          <div class="text-center">  
+            <button class="btn btn-primary" onclick="window.location.href='https://ezbc.me/2018/08/07/drug-adverse-event-entity-extraction'">Visit Blog Post</button>
+          </div>
+        </p>
+        </div>
+
+    </div><!-- /.container -->
+  </div>
+    <!-- Bootstrap core JavaScript
```
Comment (ezbc) [inline templates/index.html]: Move `<script>` calls to a separate template, `footer.html`. Add `footer.html` to the layout so that all pages load javascript.

File: templates/index.html [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+    ================================================== -->
+    <!-- Placed at the end of the document so the pages load faster -->
+    <script src="https://ajax.googleapis.com/ajax/libs/jquery/1.12.4/jquery.min.js"></script>
+    <!-- Latest compiled and minified JavaScript -->
+    <script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
+
+    <script src="https://cdnjs.cloudflare.com/ajax/libs/popper.js/1.12.9/umd/popper.min.js"></script>
+
+    <script>
```
Comment (ezbc) [inline templates/index.html]: Move javascript to a static file at `static/js/app.js`. Load the javascript file with
```html
<script src="{{url_for('static', filename='js/app.js')}}"></script>
```

File: templates/response.html [hunk 15]
```diff
@@ -0,0 +1,8 @@
+
+{% if annotated_text %}
+  <h2>Results:</h2>
+  {% for block in annotated_text %}
+    {% if block.label %}<span class="highlight highlight-{{ block.class }}" data-toggle="popover" data-placement="top" data-content="{{ block.label }}">{{ block.text }}</span>{% else %}<span>{{ block.text }}</span>{% endif %}
```
Comment (ezbc) [inline templates/response.html]: Create a separate template, `annotated_text.html` which builds a highlighted span of text.

Leave the conditional check for if the block has a label in `response.html`.

File: main_test.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
+# Copyright 2016 Google Inc. All Rights Reserved.
```
Comment (ezbc) [inline main_test.py]: Create a new `tests` folder. Move `main_test.py` to `tests`.

File: main_test.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+    client = main.app.test_client()
+    return client
+
+class Request:
+
+    def __init__(self, method, form):
+        self.method = method
+        self.form = form
+
+def test__get_entities_from_text_expecting_none():
```
Comment (ezbc) [inline main_test.py]: Add comments about the context of each test. 

What state is the app in when `_get_entities_from_text` is expecting None?

File: drugner.py [hunk 18]
```diff
@@ -0,0 +1,28 @@
+import spacy
+
+class DrugNER:
+
+    def __init__(self, model_name, model_path):
+    
+        self.model_name = model_name
+        self.model_path = model_path
+        self.model = None
```
Comment (ezbc) [inline drugner.py]: Now that we have the lazy model load implemented, add unit tests for this module.

## Steps

Create a new module for pytest in a [`tests` directory](https://github.com/ezbc/drug-portal-ner-app-code-review/pull/1/files).

Create a mock model that mimics the portions of the spacy API we use.

In each test initialize an instance of `DrugNER` with the mocked model.
